src/java/org/apache/cassandra/index/sai/disk/vector/CompactionGraph.java

```diff
@@ -250,6 +259,38 @@ public InsertionResult maybeAddVector(ByteBuffer term, int segmentRowId) throws
             int ordinal = useSyntheticOrdinals ? nextOrdinal++ : segmentRowId;
             postings = new CompactionVectorPostings(ordinal, segmentRowId);
             postingsMap.put(vector, postings);
+
+            // fine-tune the PQ if we've collected enough vectors
+            if (!pqFinetuned && postingsMap.size() >= PQ_TRAINING_SIZE)
```

Comment:
It looks like there might be a race with `pqFinetuned` where we could get two threads entering this conditional block when it is `false` and about to be `true`.

Reply:
maybeAddVector is single-threaded

Reply:
Oh, that's right, sorry, I forgot we are in compaction graph.